tac_plus.1
.TH tac_plus 8 "10 February 1995"
.SH NAME
tac_plus \- tacacs plus daemon
.SH SYNOPSIS
.B tac_plus
.B \-C\ <configfile>
[
.B \-t
] [
.B \-P
] [
.B \-g
] [
.B \-i
] [
.B \-v
] [
.B \-L
] [
.B \-p port
] [
.B \-d level
]
.SH DESCRIPTION
tac_plus listens on tcp port
.B
49 
and provides Cisco systems routers and access servers with
authentication, authorisation and accounting services.
.LP
A configuration file controls the details of authentication,
authorisation and accounting.
.LP
On startup, tac_plus creates the file
.B /etc/tac_plus.pid ,
if possible, containing its process id.
.LP
.SH ARGUMENTS and OPTIONS
.TP
.B \-C <configfile>
.IP
Specify the configuration file name. A configuration file is
.B
always required.
.TP 
.B \-P
Just parse the configuration file, echoing it to standard output while
parsing, and then exit. Used for debugging configuration file syntax.
.TP
.B \-t
Log all informational, debugging or error messages to
.B
/dev/console 
in addition to logging to syslogd. Useful for debugging.
.IP
.B
NOTE: 
messages at priority LOG_DEBUG are never logged to syslog, Use the
.B
\-t, \-d or \-g 
flags to see all messages produced by tac_plus.  These flags
should not be used in normal service.
.TP
.B \-g
Go into single threaded mode, only accepting and servicing a single
connection at a time without forking and without closing file
descriptors.  Print all messages to standard output. For debugging
only. Don't ever try to deliver normal service this way.
.TP
.B \-v
Print the current version of tac_plus to stdout and then exit.
.TP
.B \-L
Lookup the hostname of the client sending requests and use if for
logging, instead of just using its ip address.
.TP
.B \-p <port>
Use the specified port number instead of the default port
.B
49 
for incoming tcp connections. Note that this changes the name of the
pid file created by the daemon, which will append the port number to
the file name if the port is not the default one.
.TP
.B \-d <level>
Switch on debugging and write debug output into
.B
/var/tmp/tac_plus.log. 

See the definitions of debugging flags at the bottom of tac_plus.h for
available flags and their meanings.  Most flags cause extra messages
to be sent to 
.B
/var/tmp/tac_plus.log 
and also to 
.B
syslog.
.IP
.B
NOTE: 
The 
.B
\-g 
flag will cause these messages to also appear on stdout.  The
.B
\-t 
flag will cause these messages to also be written to /dev/console.
.IP
The values represent bits, so they can be added together. Currently
the following values are recognised:
.nf

Value   Meaning
8       authorisation debugging
16      authentication debugging
32      password file processing debugging
64      accounting debugging
128     config file parsing & lookup
256     packet transmission/reception
512     encryption/decryption
1024    MD5 hash algorithm debugging
2048    very low level encryption/decryption

.fi
.TP
.B \-i
Run under inetd instead of running standalone. Under inetd, the config
file is parsed every time tac_plus starts up, so this is very
inefficient if the config file is large or there are many incoming
connections. The standalone version only reads the config file once at
startup.
.IP
If the config file is small, and you don't have very frequent incoming
connections, and authentication is being done via passwd(5) files or
SKEY (which are not cached), running under inetd should be tolerable,
but still isn't recommended.
.TP
\-s
.IP
The \-s flag will cause the daemon to always reject authentication
requests which contain a minor version number of zero (SENDPASS).  You
can do this only if all your NASes are running an IOS version of 11.2
or later.
.IP
This enhances security in the event that someone discovers your
encryption key.  SENDPASS requests permits requestors to obtain chap,
pap and arap passwords from your daemon, if (and only if) they know
your encryption key.
.LP
.SH INVOKING TAC_PLUS
.LP
Tac_plus is normally invoked by root, as follows:
.LP

    # tac_plus -C <configfile>

.LP
where <configfile> is a full path to the configuration file. Tac_plus
will background itself and start listening on port 49 for incoming tcp
connections.
.LP
Tac_plus must be invoked as root to obtain privileged network socket
49 and to read the protected configuration file which may contain
confidential information such as encryption keys and cleartext
passwords.
.LP
After the port is acquired and the config file is read, root
privileges are no longer required.  You can arrange that tac_plus will
change its user and groupid to more innocuous user and group (see the
Makefile for instructions on how to do compile this) when
appropriate.
.LP
.B
NOTE:
The new user and group still needs permission to read any
passwd(5) files and S/KEY database if these are being used.
.SH CONFIGURATION FILE PERMISSIONS
.LP
It goes without saying (though I say it here) that the configuration
file should be unreadable and unwriteable by anyone except root, as it
contains passwords and keys.
.SH UPDATING THE CONFIGURATION FILE
.LP
If the daemon is sent a SIGUSR1, it will reinitialize itself,
re-reading its config file from scratch. Note that if there is an
error in the CONFIG file, the daemon will die.
.LP
.SH SYSLOG MESSAGES
.LP
tac_plus logs error messages to syslog, and informational messages to
facility LOG_LOCAL6. Debug messages are never sent to syslog.
.LP
You may wish to add a line similar to the following to your
syslog.conf file to see the informational messages logged using this
facility.
.nf

local6.info					/var/adm/messages

.fi
.LP
Note that in some versions of syslogd e.g. SunOS, this line must
contain only tabs, not spaces, and that syslogd gives very little in
the way of diagnostics when it encounters errors in the syslog.conf
file.
.fi
.SH SEE ALSO
.LP
The tac_plus User's Guide.
.SH FILES
.TP 30
.B /var/tmp/tac_plus.log
Contains debugging output when -d is in effect.
.TP
.B /etc/tac_plus.pid  or /etc/tac_plus.pid.port
contains the process id of the currently running daemon.  The port
number is appended to the filename only if the port being used is not
the default one of 49.
.SH BUGS
The configuration file syntax is too complex.


CHANGES
>Changes from Release 0.0 to release 0.1
---------------------------------------

You must now say "default attribute = permit" instead of 
default authorization = permit" when configuring service
defaults

You must now say "default svc = permit" instead of "default
authorization = permit" when configuring service defaults for a user.

When authorizing enable requests, the daemon used to prompt for a
username which it then ignored. It no longer prompts for a username.

Fix recursion issues with service and command lookups. They are now
fully recursive c.f. password lookups.

Add debugging output to password verification to provide information
about expiry processing.

Keep track of longest hash chain we create, for fine tuning. Hash all
keywords into a keyword table instead of doing linear lookup.

Update users_guide to reflect the new configuration syntax.

The convert.pl script now generates the new configuration file syntax.

Accounting code now honours the "more" flag.


Changes from Release 0.1 to release 0.2
---------------------------------------

You can now send a SIGHUP to the daemon to cause it to reinitialize
itself and re-read its CONFIG file. There is a new debugging flag
devoted to this section of the code.

Node types are now pretty-printed in debug output.

The conversion script "convert.pl" now will not print out an expires
field if it doesn't think the syntax of the field is correct. It also
now ignores blank lines in its input files.

When doing authorization, the NAS supplied attribute "cmd=" is now
correctly ignored. This would previously have caused exec
authorization to be denied.

Changes from Release 0.2 to release 0.3
---------------------------------------
Warn when not invoked as uid 0.
Improved Usage message
Add make install target

Changes from Release 0.3 to release 0.4
---------------------------------------
Add TAC_PLUS_PIDFILE to makefile per Andy Linton's suggestion.
Fix bug in authorization code (protocol field needs to be
uppercase) which prevented authorization from working.

Changes from Release 0.4 to release 0.5
---------------------------------------
Add pre and post authorization calls to shell commands.
Minor bugfixes and code cleanup
The "More" bit in accounting records is now honoured.
Fix a bug in convert.pl
Redo accounting output routines. You can now name the accounting file
in the configuration file.
Change "svc" to "service" and "proto" to "protocol".
You can use any string to name a ppp protocol, even one which doesn't yet exist.
Add PPP/LCP special case processing
Revised authorization algorithm (see user's guide)
Add hex debug flag to allow skipping hex in packet dumps.
Update user's guide to reflect changes

Changes from Release 0.4 to release 1.0
---------------------------------------
Changed format of syslog messages to make writing scripts easier
Added ability to use cleartext passwords instead of DES passwords
Updated man page to reflect the fact that we use SIGUSR1 to re-read
the config file. SIGHUP is now ignored.
Updated the users guide.

Changes from Release 1.0 to release 1.1
---------------------------------------
Release 1.1 corresponds to RCS version 1.64 of tac_plus 
(see tac_plus -v)

A typo in the Solaris section of the Makefile has been fixed.

The keyword 'des' has been introduced which must be used before all
des encrypted passwords.

The keyword 'password' has been changed to 'login', so
    password = f23sac783n
has become
    login = des f23sac783n

The convert.pl script knows about these changes.

arap and chap now require the keyword 'cleartext' in front of their
passwords.

A cleartext, per-user, global password can now be configured, which
works for login, arap and chap.

The users_guide has been updated to include a list of all A/V pairs
recognised by IOS 10.3(3) code.

Some solaris binaries have been provided as a courtesy.

Changes from Release 1.1 to release 2.0
---------------------------------------
generate_password.pl has been removed in favour of a C program
generate_passwd.c

The version number reported by tac_plus has been changed to agree with the
release number. This is why the version has jumped to 2.0

skey was broken by changes made in 1.1. These are now fixed.

Documentation has been added for the authorization AV pairs supported
by IOS releases 10.3(3) and 11.0.

Changes from Release 2.0 to release 2.1
---------------------------------------
There are now Makefile definitions for most of the major platforms.

Minor changes to remove some spurious debugging output.

A prematurely closed NAS connection will now call the authentication
function with the abort flag set, so that it can do any clean up it
requires.

syslog messages will contain the string "unknown" for usernames and
ports which are NULL, so that the messages always contain a fixed
number of fields.

The authentication code has been rearranged to better reflect the
structure of the API.

The "default user = permit" directive is still accepted but is now
deprecated in favour of "default authorization = permit".

A bug in the handling of substring AV pairs which caused the attribute
"addr" to erroneously match "addr-pool" has been fixed.

Added new files: enable.c generate_passwd.c skey_fn.c 

New #defines have been added to make it easier to port tacacs+ to new
systems.

Many more iterations are allowed before an error is declared.

Changes from Release 2.1 to release 2.2
---------------------------------------
The expiry field in the shadow file on Solaris machines is now
honored, if it exists.

Added TAC_PLUS_AUTHEN_SVC_NASI

Changes from Release 2.2 to release F3.0.13
-------------------------------------------
NEW REVISION OF THE PROTOCOL corresponding to tacacs+.spec.v1.63.ps
(which see) to increase security in the case of compromised keys.

Inbound pap logins and outbound pap password are now configurable as
separate entries for each principal.  Inbound pap logins are now

tac_plus.c
/*
 * tac_plus.c
 *
 * TACACS_PLUS daemon suitable for using on Un*x systems.
 *
 * October 1994, Lol Grant
 *
 * Copyright (c) 1994-2000 by Cisco systems, Inc.
 *
 * Permission to use, copy, modify, and distribute modified and
 * unmodified copies of this software for any purpose and without fee is
 * hereby granted, provided that (a) this copyright and permission notice
 * appear on all copies of the software and supporting documentation, (b)
 * the name of Cisco Systems, Inc. not be used in advertising or
 * publicity pertaining to distribution of the program without specific
 * prior permission, and (c) notice be given in supporting documentation
 * that use, modification, copying and distribution is by permission of
 * Cisco Systems, Inc.

 * Cisco Systems, Inc. makes no representations about the suitability
 * of this software for any purpose.  THIS SOFTWARE IS PROVIDED ``AS
 * IS'' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,
 * WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE.
*/

#include "tac_plus.h"
#include "sys/wait.h"
#include "signal.h"

static int standalone  = 1; /* running standalone (1) or under inetd (0) */
static int initialised = 0; /* data structures have been allocated */
int sendauth_only      = 0; /* don't respond to sendpass requests */
int debug              = 0; /* debugging flags */
int port               = 0; /* port we're listening on */
int console            = 0; /* write all syslog messages to console */
int parse_only         = 0; /* exit after verbose parsing */
int single             = 0; /* single thread (for debugging) */
int wtmpfd	       = 0; /* for wtmp file logging */
char *wtmpfile         = NULL;

struct timeval started_at;

struct session session;     /* session data */

static char pidfilebuf[75]; /* holds current name of the pidfile */

void start_session();

#ifndef REAPCHILD
static
#ifdef VOIDSIG
void 
#else
int
#endif /* VOIDSIG */
reapchild()
{
#ifdef UNIONWAIT
    union wait status;
#else
    int status;
#endif
    int pid;

    for (;;) {
	pid = wait3(&status, WNOHANG, 0);
	if (pid <= 0)
	    return;
	if (debug & DEBUG_FORK_FLAG)
	    report(LOG_DEBUG, "%d reaped", pid);
    }
}
#endif /* REAPCHILD */

static void
die(signum)
int signum;
{
    report(LOG_INFO, "Received signal %d, shutting down", signum);
    unlink(pidfilebuf);
    tac_exit(0);
}

static void
init()
{
    if (initialised)
	cfg_clean_config();    

    report(LOG_INFO, "Reading config");

    session.acctfile = tac_strdup("/var/tmp/acctfile");
    
    if (!session.cfgfile) {
	report(LOG_ERR, "no config file specified");
	tac_exit(1);
    }
    
    /* read the config file */
    if (cfg_read_config(session.cfgfile)) {
	report(LOG_ERR, "Parsing %s", session.cfgfile);
	tac_exit(1);
    }

    initialised++;

    report(LOG_INFO, "Version %s Initialized %d", VERSION, initialised);

}

static void
handler(signum)
int signum;
{
    report(LOG_INFO, "Received signal %d", signum);
    init();
#ifdef REARMSIGNAL
    signal(SIGUSR1, handler);
    signal(SIGHUP, handler);
#endif REARMSIGNAL
}

/*
 * Return a socket bound to an appropriate port number/address. Exits
 * the program on failure */

get_socket()
{
    int s;
    struct sockaddr_in sin;
    struct servent *sp;
    int on = 1;

    bzero((char *) &sin, sizeof(sin));

    if (port) {
	sin.sin_port = htons(port);
    } else {
	sp = getservbyname("tacacs", "tcp");
	if (sp)
	    sin.sin_port = sp->s_port;
	else {
	    report(LOG_ERR, "Cannot find socket port");
	    tac_exit(1);
	}
    }

    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = htonl(INADDR_ANY);

    s = socket(AF_INET, SOCK_STREAM, 0);

    if (s < 0) {
	console++;
	report(LOG_ERR, "get_socket: socket: %s", sys_errlist[errno]);
	tac_exit(1);
    }
#ifdef SO_REUSEADDR
	if (setsockopt(s, SOL_SOCKET, SO_REUSEADDR, (char *) &on,
		       sizeof(on)) < 0)
	    perror("setsockopt - SO_REUSEADDR");
#endif				/* SO_REUSEADDR */

    if (bind(s, (struct sockaddr *) &sin, sizeof(sin)) < 0) {
	console++;
	report(LOG_ERR, "get_socket: bind %d %s",
	       ntohs(sin.sin_port),
	       sys_errlist[errno]);
	tac_exit(1);
    }
    return (s);
}

static void
open_logfile()
{
#ifdef LOG_LOCAL6
    openlog("tac_plus", LOG_PID, LOG_LOCAL6);
#else
    openlog("tac_plus", LOG_PID);
#endif
    setlogmask(LOG_UPTO(LOG_DEBUG));
}

/*
 * main
 *
 * We will eventually be called from inetd or via the rc scripts directly
 * Parse arguments and act appropiately.
 */

main(argc, argv)
int argc;
char **argv;
{
    extern char *optarg;
    int childpid;
    int c;
    int s;
    FILE *fp;
    int lookup_peer = 0;

    debug = 0;			/* no debugging */
    standalone = 1;			/* standalone */
    single = 0;			/* single threaded */

    /* initialise global session data */
    bzero(&session, sizeof(session));
    session.peer = tac_strdup("unknown");

    open_logfile();

#ifdef TAC_PLUS_PORT
    port = TAC_PLUS_PORT;
#endif

    if (argc <= 1) {
	fprintf(stderr, "Usage: tac_plus -C <configuration file>\n");
	fprintf(stderr, "\t[ -t ] [ -P ] [ -g ] [ -p <port> ]\n");
	fprintf(stderr, "\t[ -d <debug level> ] [ -i ] [ -v ] [ -s ]\n");
	fprintf(stderr, "\t[ -l logfile ]");
#ifdef MAXSESS
	fprintf(stderr, " [ -w whologfile ]");
#endif
	fprintf(stderr, "\n");
	tac_exit(1);
    }

    while ((c = getopt(argc, argv, "td:C:ip:PgvsLl:w:u:")) != EOF)
	switch (c) {
	case 'L':		/* lookup peer names via DNS */
	    lookup_peer++;
	    break;
	case 's':		/* don't respond to sendpass */
	    sendauth_only++;
	    break;
	case 'v':		/* print version and exit */
	    version();
	    tac_exit(1);
	case 't':
	    console++;		/* log to console too */
	    break;
	case 'P':		/* Parse config file only */
	    parse_only++;
	    break;
	case 'g':		/* single threaded */
	    single++;
	    break;
	case 'p':		/* port */
	    port = atoi(optarg);
	    break;
	case 'd':		/* debug */
	    debug = atoi(optarg);
	    break;
	case 'C':		/* config file name */
	    session.cfgfile = tac_strdup(optarg);
	    break;
	case 'i':		/* stand-alone */
	    standalone = 0;
	    break;
	case 'l':		/* logfile */
	    logfile = tac_strdup(optarg);
	    break;
#ifdef MAXSESS
	case 'w':		/* wholog file */
	    wholog = tac_strdup(optarg);
	    break;
#endif
	case 'u':
	    wtmpfile = tac_strdup(optarg);
	    break;

	default:
	    fprintf(stderr, "%s: bad switch %c\n", argv[0], c);
	    tac_exit(1);
	}

    if (geteuid() != 0) {
	fprintf(stderr, "Warning, not running as uid 0\n");
	fprintf(stderr, "Tac_plus is usually run as root\n");
    }

    parser_init();

    init();

    signal(SIGUSR1, handler);
    signal(SIGHUP, handler);
    signal(SIGTERM, die);
    signal(SIGPIPE, SIG_IGN);

    if (parse_only)
	tac_exit(0);

    if (debug)
	report(LOG_DEBUG, "tac_plus server %s starting", VERSION);

    if (!standalone) {
	/* running under inetd */
	struct sockaddr_in name;
	int name_len;
	int on = 1;

	name_len = sizeof(name);

	session.sock = 0;
	if (getpeername(session.sock, (struct sockaddr *) &name, &name_len)) {
	    report(LOG_ERR, "getpeername failure %s", sys_errlist[errno]);
	} else {
	    struct hostent *hp;
	    hp = gethostbyaddr((char *) &name.sin_addr.s_addr,
			       sizeof(name.sin_addr.s_addr), AF_INET);
	    if (session.peer) {
		free(session.peer);
	    }
	    session.peer = tac_strdup(hp ? hp->h_name : 
				  (char *) inet_ntoa(name.sin_addr));
	}
#ifdef FIONBIO
	if (ioctl(session.sock, FIONBIO, &on) < 0) {
	    report(LOG_ERR, "ioctl(FIONBIO) %s", sys_errlist[errno]);
	    tac_exit(1);
	}
#endif
	start_session();
	tac_exit(0);
    }

    if (!single) {
	/* Running standalone. Background ourselves, let go of controlling tty */

#ifdef SIGTTOU
	signal(SIGTTOU, SIG_IGN);
#endif
#ifdef SIGTTIN
	signal(SIGTTIN, SIG_IGN);
#endif
#ifdef SIGTSTP
	signal(SIGTSTP, SIG_IGN);
#endif
	
	signal(SIGHUP, SIG_IGN);
    
	if ((childpid = fork()) < 0)
	    report(LOG_ERR, "Can't fork first child");
	else if (childpid > 0)
	    exit(0);		/* parent */

	if (debug)
	    report(LOG_DEBUG, "Backgrounded");

#ifndef REAPCHILD

#ifdef LINUX
	if (setpgrp() == -1)
#else /* LINUX */
	if (setpgrp(0, getpid()) == -1)
#endif /* LINUX */
	    report(LOG_ERR, "Can't change process group");
	
	c = open("/dev/tty", O_RDWR);
	if (c >= 0) {
	    ioctl(c, TIOCNOTTY, (char *) 0);
	    (void) close(c);
	}
	signal(SIGCHLD, reapchild);

#else /* REAPCHILD */

	if (setpgrp() == 1)
	    report(LOG_ERR, "Can't change process group");

	signal(SIGHUP, SIG_IGN);

	if ((childpid = fork()) < 0)
	    report(LOG_ERR, "Can't fork second child");
	else if (childpid > 0)
	    exit(0);
    
	if (debug & DEBUG_FORK_FLAG)
	    report(LOG_DEBUG, "Forked grandchild");

	signal(SIGCHLD, SIG_IGN);

#endif /* REAPCHILD */

	closelog(); /* some systems require this */

	for (c = 0; c < getdtablesize(); c++)
	    (void) close(c);

	/* make sure we can still log to syslog now we've closed everything */
	open_logfile();

    } /* ! single threaded */
    
    ostream = NULL;
    /* chdir("/"); */
    umask(0);
    errno = 0;

    s = get_socket();

#ifndef SOMAXCONN
#define SOMAXCONN 5
#endif

    if (listen(s, SOMAXCONN) < 0) {
	console++;
	report(LOG_ERR, "listen: %s", sys_errlist[errno]);
	tac_exit(1);
    }

    if (port == TAC_PLUS_PORT) {
	strcpy(pidfilebuf, TAC_PLUS_PIDFILE);
    } else {
	sprintf(pidfilebuf, "%s.%d", TAC_PLUS_PIDFILE, port);
    }

    /* write process id to pidfile */
    if ((fp = fopen(pidfilebuf, "w")) != NULL) {
	fprintf(fp, "%d\n", getpid());
	fclose(fp);
    } else 
	report(LOG_ERR, "Cannot write pid to %s %s", 
	       pidfilebuf, sys_errlist[errno]);

#ifdef TAC_PLUS_GROUPID
    if (setgid(TAC_PLUS_GROUPID))
	report(LOG_ERR, "Cannot set group id to %d %s", 
	       TAC_PLUS_GROUPID, sys_errlist[errno]);
#endif

#ifdef TAC_PLUS_USERID
    if (setuid(TAC_PLUS_USERID)) 
	report(LOG_ERR, "Cannot set user id to %d %s", 
	       TAC_PLUS_USERID, sys_errlist[errno]);
#endif

#ifdef MAXSESS
    maxsess_loginit();
#endif /* MAXSESS */

    report(LOG_DEBUG, "uid=%d euid=%d gid=%d egid=%d s=%d",
	   getuid(), geteuid(), getgid(), getegid(), s);

    for (;;) {
	int pid;
	struct sockaddr_in from;
	int from_len;
	int newsockfd;
	struct hostent *hp = NULL;

	bzero((char *) &from, sizeof(from));
	from_len = sizeof(from);

	newsockfd = accept(s, (struct sockaddr *) &from, &from_len);

	if (newsockfd < 0) {
	    if (errno == EINTR)
		continue;

	    report(LOG_ERR, "accept: %s", sys_errlist[errno]);
	    continue;
	}

	if (lookup_peer) {
	    hp = gethostbyaddr((char *) &from.sin_addr.s_addr,
			       sizeof(from.sin_addr.s_addr), AF_INET);
	}

	if (session.peer) {
	    free(session.peer);
	}
	session.peer = tac_strdup(hp ? hp->h_name : 
				  (char *) inet_ntoa(from.sin_addr));

	if (debug & DEBUG_PACKET_FLAG)
	    report(LOG_DEBUG, "session request from %s sock=%d", 
		   session.peer, newsockfd);

	if (!single) {
	    pid = fork();

	    if (pid < 0) {
		report(LOG_ERR, "fork error");
		tac_exit(1);
	    }
	} else {
	    pid = 0;
	}

	if (pid == 0) {
	    /* child */
	    if (!single)
		close(s);
	    session.sock = newsockfd;
	    start_session();
	    shutdown(session.sock, 2);
	    close(session.sock);
	    if (!single)
		tac_exit(0);
	} else {
	    if (debug & DEBUG_FORK_FLAG)
		report(LOG_DEBUG, "forked %d", pid);
	    /* parent */
	    close(newsockfd);
	}
    }
}

#ifdef GETDTABLESIZE
int 
getdtablesize()
{
    return(_NFILE);
}
#endif /* GETDTABLESIZE */

/* Make sure version number is kosher. Return 0 if it is */
int
bad_version_check(pak)
u_char *pak;
{
    HDR *hdr = (HDR *) pak;
    
    switch (hdr->type) {
    case TAC_PLUS_AUTHEN:
	/* 
	 * Let authen routines take care of more sophisticated version
	 * checking as its now a bit involved. 
	 */
	return(0);

    case TAC_PLUS_AUTHOR:
    case TAC_PLUS_ACCT:
	if (hdr->version != TAC_PLUS_VER_0) {
	    send_error_reply(hdr->type, "Illegal packet version");
	    return(1);
	}
	return(0);

    default:
	return(1);
    }
}

/*
 * Determine the packet type, read the rest of the packet data,
 * decrypt it and call the appropriate service routine.
 *
 */

void
start_session()
{
    u_char *pak, *read_packet();
    HDR *hdr;
    void authen();

    session.seq_no = 0;
    session.aborted = 0;
    session.version = 0;

    pak = read_packet();
    if (!pak) {
	return;
    }

    if (debug & DEBUG_PACKET_FLAG) {
	report(LOG_DEBUG, "validation request from %s", session.peer);
	dump_nas_pak(pak);
    }
    hdr = (HDR *) pak;

    session.session_id = ntohl(hdr->session_id);

    /* Do some version checking */
    if (bad_version_check(pak)) {
	free(pak);
	return;
    }

    switch (hdr->type) {
    case TAC_PLUS_AUTHEN:
	authen(pak);
	free(pak);
	return;

    case TAC_PLUS_AUTHOR:
	author(pak);
	free(pak);
	return;

    case TAC_PLUS_ACCT:
	accounting(pak);
	return;

    default:
	/* Note: can't send error reply if type is unknown */
	report(LOG_ERR, "Illegal type %d in received packet", hdr->type);
	free(pak);
	return;
    }
}

version()
{
    fprintf(stdout, "tac_plus version %s\n", VERSION);
#ifdef AIX
    fprintf(stdout,"AIX\n");
#endif
#ifdef ARAP_DES
    fprintf(stdout,"ARAP_DES\n");
#endif
#ifdef BSDI
    fprintf(stdout,"BSDI\n");
#endif
#ifdef CONST_SYSERRLIST
    fprintf(stdout,"CONST_SYSERRLIST\n");
#endif
#ifdef DEBUG
    fprintf(stdout,"DEBUG\n");
#endif
#ifdef DES_DEBUG
    fprintf(stdout,"DES_DEBUG\n");
#endif
#ifdef FIONBIO
    fprintf(stdout,"FIONBIO\n");
#endif
#ifdef FREEBSD
    fprintf(stdout,"FREEBSD\n");
#endif
#ifdef GETDTABLESIZE
    fprintf(stdout,"GETDTABLESIZE\n");
#endif
#ifdef HPUX
    fprintf(stdout,"HPUX\n");
#endif
#ifdef LINUX
    fprintf(stdout,"LINUX\n");
#endif
#ifdef LITTLE_ENDIAN
    fprintf(stdout,"LITTLE_ENDIAN\n");
#endif
#ifdef LOG_LOCAL6
    fprintf(stdout,"LOG_LOCAL6\n");
#endif
#ifdef MAXSESS
    fprintf(stdout,"MAXSESS\n");
#endif
#ifdef MIPS
    fprintf(stdout,"MIPS\n");
#endif
#ifdef NEED_BZERO
    fprintf(stdout,"NEED_BZERO\n");
#endif
#ifdef NETBSD
    fprintf(stdout,"NETBSD\n");
#endif
#ifdef NO_PWAGE
    fprintf(stdout,"NO_PWAGE\n");
#endif
#ifdef REAPCHILD
    fprintf(stdout,"REAPCHILD\n");
#endif
#ifdef REARMSIGNAL
    fprintf(stdout,"REARMSIGNAL\n");
#endif
#ifdef SHADOW_PASSWORDS
    fprintf(stdout,"SHADOW_PASSWORDS\n");
#endif
#ifdef SIGTSTP
    fprintf(stdout,"SIGTSTP\n");
#endif
#ifdef SIGTTIN
    fprintf(stdout,"SIGTTIN\n");
#endif
#ifdef SIGTTOU
    fprintf(stdout,"SIGTTOU\n");
#endif
#ifdef SKEY
    fprintf(stdout,"SKEY\n");
#endif
#ifdef SOLARIS
    fprintf(stdout,"SOLARIS\n");
#endif
#ifdef SO_REUSEADDR
    fprintf(stdout,"SO_REUSEADDR\n");
#endif
#ifdef STDLIB_MALLOC
    fprintf(stdout,"STDLIB_MALLOC\n");
#endif
#ifdef STRCSPN
    fprintf(stdout,"STRCSPN\n");
#endif
#ifdef SYSLOG_IN_SYS
    fprintf(stdout,"SYSLOG_IN_SYS\n");
#endif
#ifdef SYSV
    fprintf(stdout,"SYSV\n");
#endif
#ifdef TAC_PLUS_GROUPID
    fprintf(stdout,"TAC_PLUS_GROUPID\n");
#endif
#ifdef TAC_PLUS_PORT
    fprintf(stdout,"TAC_PLUS_PORT\n");
#endif
#ifdef TAC_PLUS_USERID
    fprintf(stdout,"TAC_PLUS_USERID\n");
#endif
#ifdef TRACE
    fprintf(stdout,"TRACE\n");
#endif
#ifdef UNIONWAIT
    fprintf(stdout,"UNIONWAIT\n");
#endif
#ifdef VOIDSIG
    fprintf(stdout,"VOIDSIG\n");
#endif
#ifdef _BSD1
    fprintf(stdout,"_BSD1\n");
#endif
#ifdef _BSD_INCLUDES
    fprintf(stdout,"_BSD_INCLUDES\n");
#endif
#ifdef __STDC__
    fprintf(stdout,"__STDC__\n");
#endif
}
